### ISSUE TITLE
super.method call incorrectly translated if class name is shadowed by local variable

### ISSUE THREAD

**Issue by yuryturing** [babel/babel#10693]
## Bug Report

**Current Behavior**
Super method call is incorrectly translated when there exists a variable shadowing class variable name. 

```js
    _createClass(_a, [{
      key: "b",
      value: function b() {
        // the next line throws "Cannot read property 'prototype' of undefined"
        var _a = _get(_getPrototypeOf(_a.prototype), "b", this).call(this);

        return _a + 1;
      }
    }]);
```
**Input Code**
[Repl](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=PTAEBcAsFNQYwPYBNYEsDOoDm0B20AnAQ3GiVACMBPCdOUAd1SlGnQDYIiCdwAoRLnThQARQCuqOAGtQAXlAAKAA7c84AJTyAfKADeoAG7dQAfSIBuUAWjhxBXGaLz4AGyLpM0AB6lcSTFUbXBE9PlAIyIpFLTDI-NAQCBhQfF9QV1R8ZIIEBkwAIgBhIlxcBBEbInJlXOVCcBoAclqKiqp6ptAEADNQcX9oHqyyAvCEiMFhJxd0cXqCADpojQtxiZs7BxmAalAARjWJgF9x44AacfNFiSlZBVuZS4jzNeO1gXdPUABBfXGVv94pt7I59mc-KcBAghCIAEIuR7SRQ_DR8aGwygufAMUBwmIY9AIVzQRauBBYRQUZYxDRAA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=es2015&prettier=false&targets=&version=7.7.3&externalPlugins=)

```js
// the code is generated by tsc with es6 target
const Quick = (parent) => { var _a; return _a = class extends parent {
      b() {
        // the next line throws "Cannot read property 'prototype' of undefined"
        const _a = super.b();
        return _a + 1;
      }
  },
  _a.Quick = Quick,
  _a;
};

class A {
  b() {
    return 1
  }
}

const B = Quick(A)

const b = new B()
console.log(b.b())
```

**Expected behavior/code**
I expect this code to output `2` in console when executed.

The shadowing variable got to be renamed if there exists a `super` call in the method body.

```js
    _createClass(_a, [{
      key: "b",
      value: function b() {
        // local variable is renamed to not shadow generated constructor function name
        var _b = _get(_getPrototypeOf(_a.prototype), "b", this).call(this);

        return _b + 1;
      }
    }]);
```

```js


```

**Babel Configuration (.babelrc, package.json, cli command)**

```json
{
  "presets": [
    [
      "@babel/preset-env"
    ]
  ]
}
```

**Environment**
- Babel version(s): v7.7.3
- Node/npm version: Node 12.3.0 / npm 6
- OS: Windows 10
- Monorepo: no
- How you are using Babel: loader

**Possible Solution**

**Additional context/Screenshots**
Add any other context about the problem here. If applicable, add screenshots to help explain.

**Comment by radekmie** [comment]
To get some traction - here is a smaller reproduction, including the TypeScript input:

```ts
const X = Y => class extends Y {
  static X = X;
  f() {
    const { x, ...xs } = super.f();
  }
};
```

```js
// TypeScript 3.7.2, target ES2015.
"use strict";
const X = Y => { var _a; return _a = class extends Y {
        f() {
            const _a = super.f(), { x } = _a, xs = __rest(_a, ["x"]);
        }
    },
    _a.X = X,
    _a; };
```

```js
// Babel 7.7.3, preset es2015.
var X = function X(Y) {
  var _a;

  return _a =
  /*#__PURE__*/
  function (_Y) {
    _inherits(_a, _Y);

    function _a() {
      _classCallCheck(this, _a);

      return _possibleConstructorReturn(this, _getPrototypeOf(_a).apply(this, arguments));
    }

    _createClass(_a, [{
      key: "f",
      value: function f() {
        var _a = _get(_getPrototypeOf(_a.prototype), "f", this).call(this),
            x = _a.x,
            xs = __rest(_a, ["x"]);
      }
    }]);

    return _a;
  }(Y), _a.X = X, _a;
};
```

**Comment by catrope** [comment]
I also ran into this issue when not using TypeScript, but just using terser. When terser minifies ES6 code, it sometimes causes method parameters to have the same name as the class they're inside of. This is perfectly valid in ES6, but fails after transformation to ES5 because of this bug. For example:

Output from terser, input into Babel:
```js
class a {
    foo(e, t, n, r, s) {
        return 'x';
    }
}

class r extends a {
    foo(e, t, n, r=false, s) {
        return super.foo(e, t, n, r, s);
    }
}
```

Output from Babel:
```js
// (function declarations of _get, _createClass, etc omitted)
var a = /*#__PURE__*/function () {
  function a() {
    _classCallCheck(this, a);
  }

  _createClass(a, [{
    key: "foo",
    value: function foo(e, t, n, r, s) {
      return 'x';
    }
  }]);

  return a;
}();

var r = /*#__PURE__*/function (_a) {
  _inherits(r, _a);

  var _super = _createSuper(r);

  function r() {
    _classCallCheck(this, r);

    return _super.apply(this, arguments);
  }

  _createClass(r, [{
    key: "foo",
    value: function foo(e, t, n) {
      var r = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
      var s = arguments.length > 4 ? arguments[4] : undefined;
      return _get(_getPrototypeOf(r.prototype), "foo", this).call(this, e, t, n, r, s);
    }
  }]);

  return r;
}(a);
```
When `r.foo()` is called, this code fails with `TypeError: Cannot convert undefined or null to object` in `_getPrototypeOf()`, because `r.prototype` thinks it's using the `r` in the outer scope (which is a class) but is actually using the `r` in the inner scope (which is a boolean).

**Comment by catrope** [comment]
This happens in real life when using prettier v2.2.1 (the latest version) from npm. In `node_modules/prettier/parser-html.js`, the following terser-minified code appears (newlines and spaces added for readability):
```js
class r extends bn.Parser{
    constructor(){
        super(En.getHtmlTagDefinition)
    }
    parse(e,t,n,r=!1,s){
        return super.parse(e,t,n,r,s)
    }
}
```
This has exactly the same shape in the example above, and so when transpiled to ES5 it fails as described. Storybook uses prettier for its dynamic source code generation feature, and it uses Babel to transform the ES6 code that ships with prettier (and some other npm modules) to ES5, resulting in dynamic source code generation being broken: https://github.com/storybookjs/storybook/issues/13721

**Comment by nicolo-ribaudo** [comment]
@catrope Which version of `@babel/plugin-transform-classes` do you have? With the last version I get this output:
```js
let r = /*#__PURE__*/function (_bn$Parser) {
  _inherits(r, _bn$Parser);

  var _super = _createSuper(r);

  function r() {
    _classCallCheck(this, r);

    return _super.call(this, En.getHtmlTagDefinition);
  }

  _createClass(r, [{
    key: "parse",
    value: function parse(e, t, n, _r = !1, s) {
      return _get(_getPrototypeOf(r.prototype), "parse", this).call(this, e, t, n, _r, s);
    }
  }]);

  return r;
}(bn.Parser);
```
The parameter has been renamed to `r`.

**Comment by nicolo-ribaudo** [comment]
Oh this has been fixed in 7.12.13: https://github.com/babel/babel/commit/cba64f9a09c2aebea65b587b31864758f3726d66
